### PR TITLE
🎉 Source S3: add json parser

### PR DESCRIPTION
**Pull request by Tutuchan**
## What
Add a JSON file parser for the S3 source (see #5103)

## How
Use the pandas JSON parser to allow syncing JSON files from S3. We cannot use PyArrow to do this as it does not allow streaming for JSON files, unlike CSV and Parquet.
Only line-delimited format is supported in this PR.

## 🚨 User Impact 🚨
No breaking changes

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

Unit tests are passing, I'm only including the final pytest result because I can't fit the whole results into one screenshot.

image

I'm having a few issues for acceptance tests and docs:

- my company uses temporary credentials to connect to AWS in eu-west-1 so I had to update a few files locally to pass the session tokens instead of only access and secret keys and change the region (side note: ideally, in production, we would strongly prefer not using keys but instance profiles and I know there are a few PRs (#5942, #5282) about this topic so I'll head over to them to see if I can help)
- I'm running acceptance tests with `python -m pytest integration_tests -p integration_tests.acceptance` and some of the existing tests pass but not others. I think it's mostly the custom server ones. I have a Minio container running with the configuration described in [the minio config](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-s3/integration_tests/config_minio.json) but I think I'm doing something wrong
image

- I added JSON test files but the prettier pre-commit refuses them as they are line-delimited JSON so I can't commit them. Is it okay to exclude them in the pre-commit configuration ?
- I'm also not sure what docs to update.

Any help and/or guidance would be greatly appreciated.

Best,
Pierre

---
This change is [Reviewable](https://reviewable.io/reviews/airbytehq/airbyte/9117)

### CODE REVIEW

**Comment by alafanechere** [comment]
Thanks, @Tutuchan! We appreciate your contribution. I'll go for a review this week.

> I'm running acceptance tests with python -m pytest integration_tests -p integration_tests.acceptance and some of the existing tests pass but not others. I think it's mostly the custom server ones. I have a Minio container running with the configuration described in the minio config but I think I'm doing something wrong.

I'll run the acceptance in our CI and share the results. 

> I added JSON test files but the prettier pre-commit refuses them as they are line-delimited JSON so I can't commit them. Is it okay to exclude them in the pre-commit configuration?

Yes, please edit the pre-commit config or commit this file with `git commit --no-verify`.

> I'm also not sure what docs to update.


Please bump the connector version and edit the changelog accordingly (in `docs/integrations/sources/s3.md`):
* `airbyte-integrations/connectors/source-s3/Dockerfile` bump `LABEL io.airbyte.version` to `0.1.8`
* `airbyte-config/init/src/main/resources/seed/source_definitions.yaml` bump `dockerImageTag` to `0.1.8`
* `airbyte-config/init/src/main/resources/config/STANDARD_SOURCE_DEFINITION/69589781-7828-43c5-9f63-8925b1c1ccc2.json` bump `dockerImageTag` to `0.1.8`

File: airbyte-integrations/connectors/source-s3/integration_tests/json_expected_records.txt [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{"stream": "test", "data": {"id": 1, "name": "PVdhmjb1", "valid": false, "_ab_additional_properties": {}, "_ab_source_file_last_modified": "2021-07-25T15:33:04+0000", "_ab_source_file_url": "simple_test.csv"}, "emitted_at": 1627227468000}
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/integration_tests/json_expected_records.txt:1]: You can use the `.jsonl` extension.

File: airbyte-integrations/connectors/source-s3/integration_tests/spec.json [hunk 2]
```diff
@@ -132,12 +133,75 @@
                 "type": "integer"
               }
             }
+          },
+          {
+            "title": "json",
+            "description": "This connector utilises <a href=\"https://pandas.pydata.org/docs/reference/api/pandas.io.json.read_json.html\" target=\"_blank\">Pandas</a> for JSON parsing.",
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/integration_tests/spec.json:139]:
```suggestion
            "description": "This connector uses <a href=\"https://pandas.pydata.org/docs/reference/api/pandas.io.json.read_json.html\" target=\"_blank\">Pandas</a> for JSON parsing.",
```

File: airbyte-integrations/connectors/source-s3/integration_tests/spec.json [hunk 3]
```diff
@@ -132,12 +133,75 @@
...
+              "keep_default_dates": {
+                "title": "Keep Default Dates",
+                "description": "If parsing dates (convert_dates is not False), then try to parse the default datelike columns. A column label is datelike if one of these conditions is verified:\n            * it ends with '_at',\n            * it ends with '_time',\n            * it begins with 'timestamp',\n            * it is 'modified',\n            * it is 'date'\n            ",
+                "default": true,
+                "type": "boolean"
+              },
+              "lines": {
+                "title": "Lines",
+                "description": "Read the file as a json object per line.",
+                "default": true,
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/integration_tests/spec.json:171]: How does the parser behave if `false`? 
I did not find any test file with has a multiline JSON object.

File: airbyte-integrations/connectors/source-s3/unit_tests/sample_files/json/test_file_1.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{"id": 1, "name": "PVdhmjb1", "valid": false, "code": 12, "degrees": -31.3, "birthday": "2021-07-14", "last_seen": "2021-07-14 15:30:09.224125"}
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/unit_tests/sample_files/json/test_file_1.json:1]: You could use `.jsonl` extension for these test files.

File: airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    # standard types
+    "string": ("string", None),
+    "boolean": ("boolean", None),
+    "number": ("number", None),
+    "integer": ("integer", None),
+    # supported by Pandas type
+    "datetime": ("string", lambda v: v.isoformat()),
+    "object": ("string", lambda v: v.isoformat()),
+    "duration": ("string", lambda v: v.isoformat()),
+    "any": ("string", lambda v: v.isoformat()),
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py:22]: Are you sure this type support iso format conversion 🤔 ? 
```suggestion
    "object": ("string", lambda v: v.isoformat()),
    "duration": ("string", lambda v: v.isoformat()),
    "any": ("string", lambda v: v.isoformat()),
```

File: airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            "encoding": self._format.get("encoding", "utf8"),
+        }
+
+    def _parse_options(self):
+        """
+        https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#reading-json
+        """
+        schema = self.json_schema_to_dtype_schema(self._master_schema) if self._master_schema is not None else None
+        return {
+            **{"dtype": schema},
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py:49]:
```suggestion
            "dtype": schema,
```

File: airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+            json_type, _ = JSON_TYPES[needed_table_schema_type]
+            return json_type
+
+    @staticmethod
+    def convert_field_data(table_schema_type: str, field_value: Any) -> Any:
+        """Converts not JSON format to JSON one"""
+        if field_value is None:
+            return None
+        if table_schema_type in JSON_TYPES:
+            _, func = JSON_TYPES[table_schema_type]
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py:73]:
```suggestion
            _, conversion_func = JSON_TYPES[table_schema_type]
```

File: airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+    @staticmethod
+    def parse_field_type(needed_table_schema_type: str) -> str:
+        """
+        https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#table-schema
+        Convert Table Schema types to Json types
+        """
+        if needed_table_schema_type not in JSON_TYPES:
+            raise TypeError(f"incorrect Table Schema type: {needed_table_schema_type}")
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py:62]:
```suggestion
            raise TypeError(f"Incorrect Table Schema type: {needed_table_schema_type}")
```

File: airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+    @staticmethod
+    def convert_field_data(table_schema_type: str, field_value: Any) -> Any:
+        """Converts not JSON format to JSON one"""
+        if field_value is None:
+            return None
+        if table_schema_type in JSON_TYPES:
+            _, func = JSON_TYPES[table_schema_type]
+            return func(field_value) if func else field_value
+        raise TypeError(f"unsupported field type: {table_schema_type}, value: {field_value}")
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py:75]:
```suggestion
        raise TypeError(f"Unsupported field type: {table_schema_type}, value: {field_value}")
```

File: airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+
+    def get_inferred_schema(self, file: Union[TextIO, BinaryIO]) -> dict:
+        """
+        https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#reading-json
+        """
+        reader = read_json(file, **self._read_options(), **self._parse_options())
+
+        schema = build_table_schema(reader.read())
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py:83]: nit
```suggestion
        reader = pd.read_json(file, **self._read_options(), **self._parse_options())

        schema = pd.build_table_schema(reader.read())
```

File: airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+        # remove the first field which is the index
+        schema_fields = schema["fields"][1:]
+        schema_dict = {field["name"]: self.parse_field_type(field["type"]) for field in schema_fields}
+
+        if not schema_dict:
+            # pandas can parse empty JSON files but a connector can't generate dynamic schema
+            raise OSError("empty JSON file")
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py:90]:
```suggestion
            raise OSError("Empty JSON file")
```

File: airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        )
+
+        is_empty = True
+        for rows in streaming_reader:
+            if len(rows) > 0:
+                is_empty = False
+            for row in rows.to_dict(orient="records"):
+                yield row
+        if is_empty:
+            raise OSError("empty JSON file")
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py:112]:
```suggestion
            raise OSError("Empty JSON file")
```

File: airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+        streaming_reader = read_json(
+            file,
+            **self._read_options(),
+            **self._parse_options(),
+        )
+
+        is_empty = True
+        for rows in streaming_reader:
+            if len(rows) > 0:
+                is_empty = False
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py:108]: nit:
```suggestion
            if is_empty and len(rows) > 0:
                is_empty = False
```

File: airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_spec.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
+#
+# Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+#
+from typing import List, Optional, Union
+
+from pydantic import BaseModel, Field
+
+
+class JsonFormat(BaseModel):
+    'This connector utilises <a href="https://pandas.pydata.org/docs/reference/api/pandas.io.json.read_json.html" target="_blank">Pandas</a> for JSON parsing.'
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_spec.py:10]:
```suggestion
    'This connector uses <a href="https://pandas.pydata.org/docs/reference/api/pandas.io.json.read_json.html" target="_blank">Pandas</a> for JSON parsing.'
```

File: airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_spec.py [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+    class Config:
+        title = "json"
+
+    filetype: str = Field(Config.title, const=True)
+
+    orient: str = Field(
+        default="columns",
+        description="Indication of expected JSON string format. Compatible JSON strings can be produced by to_json() with a corresponding orient value. Allowed values are {'split','records','index', 'columns','values', 'table'}",
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_spec.py:19]:
```suggestion
        description="The expected JSON string format. Details can be found in the <a href="https://pandas.pydata.org/docs/reference/api/pandas.read_json.html">Pandas documentation</a>",
        examples=['split','records','index', 'columns','values', 'table']"
```

File: airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_spec.py [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    lines: bool = Field(default=True, description="Read the file as a json object per line.")
+
+    chunk_size: int = Field(
+        default=100,
+        description="If lines is True, returns a JsonReader iterator to read batches of `chunk_size` lines instead of the whole file at once.",
+    )
+
+    compression: str = Field(
+        default="infer",
+        description='For on-the-fly decompression of on-disk data. If "infer", then use gzip, bz2, zip or xz if path_or_buf is a string ending in ".gz", ".bz2", ".zip", or "xz", respectively, and no decompression otherwise. If using "zip", the ZIP file must contain only one data file to be read in. Set to None for no decompression.',
```
Comment (alafanechere) [inline airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_spec.py:46]:
```suggestion
        description='For on-the-fly decompression of on-disk data. If "infer", then use gzip, bz2, zip or xz if path_or_buf is a string ending in ".gz", ".bz2", ".zip", or "xz", respectively, and no decompression otherwise. If using "zip", the ZIP file must contain only one data file to be read in. Set to None for no decompression.',
        examples=[".gz", ".bz2", ".zip", "xz"]
```

**Review by alafanechere — CHANGES_REQUESTED** [review]
Thanks for this top-quality addition. I made a minor suggestion and have a small concern: Will the parser work for files that are "classic" JSON files and not JSON lines files? If it's not, I'd suggest that you explicitly use the term `JSONL` in this PR.  If it only supports JSONL I'm ok with this and "classic" JSON can be a later addition.

I request changes because:
* The acceptance tests are not passing in our CI (you can check the failure [here], I need to add the `json_config.json` secret file to our CI)(https://github.com/airbytehq/airbyte/pull/9160#issuecomment-1002201054)
* Please bump the connector version (check my first comment on this PR for details)
* A small conflict on the `.gitignore` needs to be resolved.

**Comment by Tutuchan** [comment]
Thank you for the detailed review and comments @alafanechere ! I'll try and address them as soon as possible.

File: airbyte-integrations/connectors/source-s3/integration_tests/spec.json (hunk 3 above)
Comment (Tutuchan) [reply airbyte-integrations/connectors/source-s3/integration_tests/spec.json:171]: You're right, I'll test this and remove the parameter if it's not applicable. In that case, I'll make sure to update all mentions of JSON to JSONL in the docs.

Comment (Tutuchan) [reply airbyte-integrations/connectors/source-s3/integration_tests/spec.json:171]: I added a test for this and it now supports multiline JSON. However, it reads the whole file in memory at once. I added a disclaimer in the docs.

File: airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py (hunk 5 above)
Comment (Tutuchan) [reply airbyte-integrations/connectors/source-s3/source_s3/source_files_abstract/formats/json_parser.py:22]: Thanks for pointing that out. I realized I'm not actually calling this function. I may have to refactor some stuff to take it into account. I'll check the conversion methods at the same time.

**Comment by alafanechere** [comment]
@Tutuchan thanks for the changes, let me know when you expect a second review from my side.

**Comment by Tutuchan** [comment]
@alafanechere I only have a couple things to check before we can go for a second round:

- do I need to create more integration test files and update the integration tests to take them into account ?
- for line-delimited JSON, the schema can be different from one record to another, which is a deviation from CSV and Parquet files. If the schema is not specified by the user, it is computed for each batch from all the records in the batch but it can still be different from one batch to another, leading to results from `AbstractFileParser.stream_records` that can look like this:
```
[
  {
    "id": 1,
    "name": "PVdhmjb1",
    "valid": false,
    "code": 12,
    "degrees": -31.3,
    "birthday": "2021-07-14",
    "last_seen": "2021-07-14 15:30:09.224125",
    "creation_date": null
  },
  {
    "id": 2,
    "name": "j4DyXTS7",
    "valid": true,
    "code": -8,
    "degrees": 41.6,
    "birthday": "2021-07-14",
    "last_seen": "2021-07-14 15:30:09.224383",
    "creation_date": "2021-07-01 15:30:09.224383"
  },
  {
    "id": 3,
    "name": "v0w8fTME",
    "valid": false,
    "code": 7,
    "degrees": -27.5,
    "birthday": "2021-07-14",
    "last_seen": "2021-07-14 15:30:09.224527"
  }
]
```
The 3rd record does not have a `creation_date` key. Will this cause issues downstream ? If so, I should make it clear in the docs that the schema should be set explicitly by the user in case of records with varying schema. What do you think ?

**Comment by alafanechere** [comment]
@Tutuchan 

> do I need to create more integration test files and update the integration tests to take them into account ?

Yes, feel free to add any test you think is helpful to validate usual cases and edge cases scenarios.

> for line-delimited JSON, the schema can be different from one record to another, which is a deviation from CSV and Parquet files. If the schema is not specified by the user, it is computed for each batch from all the records in the batch but it can still be different from one batch to another, leading to results from AbstractFileParser.stream_records that can look like this. The 3rd record does not have a creation_date key. Will this cause issues downstream ? If so, I should make it clear in the docs that the schema should be set explicitly by the user in case of records with varying schema. What do you think ?

We should explicitly mention from which file the schema is inferred. The catalog and its schema must remain identical for each batch. You could either require the user to define the schema in the config or explicitly log from which file the schema is inferred. Does it answer your question?

**Comment by alafanechere** [comment]
Hi @Tutuchan, an other PR was merged for this connector, leading to some conflicts with `master`. Could you please fix them?  Could you also let me know when you need an additional review from my side?

**Comment by alafanechere** [comment]
Hi @Tutuchan, I fixed the conflicts myself.

**Comment by Tutuchan** [comment]
Hi @alafanechere , I'm really sorry, I don't have much bandwidth at the moment, I'll try to find the time by the end of the week or next week at the latest.

**Comment by Tutuchan** [comment]
> Yes, feel free to add any test you think is helpful to validate usual cases and edge cases scenarios.

Ok, I'll add some more tests then.

> We should explicitly mention from which file the schema is inferred. The catalog and its schema must remain identical for each batch. You could either require the user to define the schema in the config or explicitly log from which file the schema is inferred. Does it answer your question?

By _each batch_, do you mean each Airbyte sync job or each batch of the parser ? In any case, I'll add a strong note for this in the docs.

**Comment by alafanechere** [comment]
> Hi @alafanechere, I'm really sorry, I don't have much bandwidth at the moment, I'll try to find the time by the end of the week or next week at the latest. Thanks for fixing the conflicts !

No problem! Just let me know if you are blocked and want our team to take over. 

> By each batch, do you mean each Airbyte sync job or each batch of the parser ? In any case, I'll add a strong note for this in the docs.
I mean the catalog should never change, so both on each sync or each "micro" batch created by the parser.

**Comment by alafanechere** [comment]
Hi @Tutuchan we closed this PR as we did not receive updates following our latest interactions. You are welcome to re-open it as soon as you want. We’ll be happy to give additional reviews. Do not hesitate to reach our team on Slack if you need support finishing this work.
